coordinator/meta_executor.go

```diff
@@ -64,45 +63,43 @@ func (m *MetaExecutor) Close() error {
 	return nil
 }
 
-func (m *MetaExecutor) EachDBNodes(database string, fn func(nodeID uint64, pts []uint32, hasError *bool) error) error {
-	_, err := m.MetaClient.Database(database)
-	if err != nil {
+func (m *MetaExecutor) EachDBNodes(database string, fn func(nodeID uint64, pts []uint32) error) error {
+	if _, err := m.MetaClient.Database(database); err != nil {
 		return err
 	}
 
 	start := time.Now()
-	//var dbPtInfo []meta2.PtInfo
-	var wg sync.WaitGroup
 
-retryExecute:
 	for {
-		if time.Since(start).Seconds() >= 30*time.Second.Seconds() {
-			break retryExecute
-		}
 		// write-available-first: skip offline pts and tolerate data lost
-		var nodePtMap map[uint64][]uint32
-		nodePtMap, err = m.MetaClient.GetNodePtsMap(database)
+		nodePtMap, err := m.MetaClient.GetNodePtsMap(database)
 		if err != nil || len(nodePtMap) == 0 {
-			break retryExecute
+			return err
 		}
-		hasErr := false
-		wg.Add(len(nodePtMap))
-		for nodeId := range nodePtMap {
+
+		chErr := make(chan error, len(nodePtMap))
+
+		for id, pts := range nodePtMap {
 			go func(nodeID uint64, pts []uint32) {
-				errR := fn(nodeID, pts, &hasErr)
-				if errR != nil {
-					err = errR
-				}
-				wg.Done()
-			}(nodeId, nodePtMap[nodeId])
+				chErr <- fn(nodeID, pts)
+			}(id, pts)
+		}
+
+		retryable := true
+		for i := 0; i < len(nodePtMap); i++ {
+			if e := <-chErr; e != nil && retryable {
+				retryable = IsRetryErrorForPtView(e)
+				err = e
+			}
 		}
-		wg.Wait()
-		if err == nil || !IsRetryErrorForPtView(err) {
-			break retryExecute
+		if err == nil || !retryable {
```

Comment:
This code may need to be changed to:
```
if err == nil {
	return nil
}

// no need to retry or timed out
if !retryable || time.Since(start) >= 30*time.Second {
	return err
}

// retry
time.Sleep(100 * time.Millisecond)
```

Reply:
I considered this but didn't make the change as I think keep the original behavior (checkout imeout after sleep) is a little better: `time.Sleep` sleeps at least the required duration and could be longer (and much longer in some rare cases), if this happens, retry may occurs after the timeout.

However, I don't have strong opinion on this and could make the change if you think it is better.

Reply:
ok

Reply:
ok